### ISSUE TITLE
Issue with @extend, placeholders and @use

### ISSUE THREAD

**Issue by gfellerph** [sass/dart-sass#1042]
We're refactoring a large codebase from @import to @use and I noticed some strange errors with placeholders and placeholders extending placeholders. To illustrate the behaviour, I created a simplified repo:

https://github.com/tuelsch/sass-playground

The interesting file is [src/placeholders.scss](https://github.com/tuelsch/sass-playground/blob/master/src/placeholders.scss). As it is, the build fails (with npm start, the first build does not print the error to the console, but rather the out.css at root). But as soon as you comment out the %whatever placeholder, the build succeeds. What's even stranger, if you just move the declaration above the %placeholder declaration, the build succeeds as well.

I don't get it. Is this a bug with sass or am I not supposed to use %placeholders in this specific way?

**Comment by cpof-tea** [comment]
It isn't mentioned in the docs (but should be), but it [was mentioned in the blog](https://sass-lang.com/blog/the-module-system-is-launched#use-the-heart-of-the-module-system).

> Note that placeholder selectors are _not_ namespaced, but they _do_ respect privacy.

So, @Awjin, it doesn't seem like a bug.

**Comment by Awjin** [comment]
This isn't a bug with placeholders, but rather how sibling extenders interact. Here is a minimal reproduction:

```
<===> parent.scss
parent { a: 1 }

<===> child.scss
@use 'parent';
child { @extend parent }
sibling { @extend child }

<===> index.scss
@use 'child';
a { @extend child }
```

But yes, in general, it is recommended to define your shared styles as mixins.

**Comment by Awjin** [comment]
More context on the bug:

Here is the expected output:

```
parent, child, sibling, index {
  a: 1;
}

child, sibling, index {
  b: 1;
}

index {
  c: 1;
}
```

But instead, it causes the following error:

```
Error: The target selector was not found.
Use "@extend parent !optional" to avoid this error.
  ╷
4 │   @extend parent;
  │   ^^^^^^^^^^^^^^
  ╵
  module-extends/child.scss 4:3  root stylesheet
```

And strangely, swapping the order of the child and sibling selectors gives the correct output.

**Comment by tuelsch** [comment]
> And strangely, swapping the order of the child and sibling selectors gives the correct output.

That's the confusing part.

**Comment by tschroeter** [comment]
Hello,

'sass' is excellent and "@extend" very useful! 

But I have to confirm that "%placeholder" in combination with "@extend" and "@use" do not work together.

Without using modules and "@use" the following code generates a more specific ".card.grid" selector. But by using modules especially this selector is no longer generated. 

Only if I additionally write ".padded {@extend %padded;} a ".grid.padded" selector will be generated.

An "@mixin" would not be sufficient for the describes use case. I would like to express: "any element which is of type 'padded' and 'grid' resets the negative margin". 

What do you think about it ?

`
%padded {
  padding: 8px;
}

%card {
  @extend %padded;
  
  background     : #fff;
  border-radius  : 5px;
}

%grid {
  margin-left    : -8px;  
}

%padded%grid {
  margin         : 0;
}

.card {
  @extend %card;
}

.grid {
  @extend %grid;
}
`

**Comment by Awjin** [comment]
> I would like to express: "any element which is of type padded and grid resets the negative margin".

This sounds like you need composition of styles, rather than a semantic relationship between `padding` and `grid`. In this case, mixins are a great choice. For example:

```scss
@mixin padding {
  padding: 8px;
}

@mixin card {
  background: #fff;
  border-radius: 5px;
}

@mixin grid {
  margin-left: -8px;
}

.card {
  @include padding;
  @include card;
}

.grid {
  @include grid;
}

.card.grid {
  margin: 0;
}
```

**Comment by tschroeter** [comment]
You're right, it is not a bug! The %placeholder runs in the global namespace.

Of course, the mentioned example also works with composition. The advantage using @extend: for many composed styles it means less writing effort.

Using @import is already deprecated. Do you know if placeholders will be removed in the future releases?

**Comment by tschroeter** [comment]
@Awjin Thanks for your reply. You're right, it is not a bug! 

The %placeholder runs in the global namespace.

Of course, the mentioned example also works with composition. The advantage using @extend: 
for many composed styles it means less writing effort.

Using @import is already deprecated. Do you know if placeholders will be removed in the future releases?

**Comment by Awjin** [comment]
> The %placeholder runs in the global namespace.
>
> Of course, the mentioned example also works with composition. The advantage using @extend:
for many composed styles it means less writing effort.

That is a good point. There is often a trade-off between verboseness and maintainability.

> Using @import is already deprecated. Do you know if placeholders will be removed in the future releases?

Although the new module system is recommended, `@import` is [not deprecated yet](https://sass-lang.com/blog/the-module-system-is-launched#future-plans). And although mixins are often recommended, placeholders are not planned for removal.

**Comment by funyapu** [comment]
Hi,
I understand things among `@extend `became so complicated with `@use`, but reporting another case here .
Got duplicate css blocks for `@extend`, and the result is hard to conjecture how and why it became like that.
Any idea? 

Version: sass@1.26.10, node v14.2.0

## Summary
Using `%placeholder` in `utils/grid` and `layout/content,header,footer`.
Forwarded `utils` and `layout` in `index.scss` and built it.

### Result
- Printed `%placeholder` rule for  3 times.
- Extended only once in `utils` but got 2 blocks for it.
- Output for `utils` contains a class for `layout/content`.
- Output for `layout/*` contains class for `utils/`.
```
--output.css--
.content, .content.-width-full header, .u-container { ...rule } // in head of "util/"
.content, .content.-width-full header, .u-container { ...rule } // in end of "util/"
.content, .content.-width-full header, .u-container, .header, .footer  { ...rule } // in head of "layout/*"
```

## Code
``` 
--index.scss--
@forward "some-other-file";
@forward "utils";
@forward "layout";
```
```
--extend.scss--
%wrapper {
  max-width: 960px;
  margin: 0 auto;
  padding: 0 8px;
}
```

```
--layout/header.scss--
@use "extend";

.header {
  @extend %wrapper;
}
```
```
--layout/content.scss--
@use "extend";

.content {
  @extend %wrapper;
}
.content.-width-full header {
  @extend %wrapper;
}
```
```
--layout/footer.scss--
@use "extend";

.footer {
  @extend %wrapper;
}
```
```
--utils/grid.scss--
@use "extend";

.u-container {
  @extend %wrapper;
}
```

**Comment by Awjin** [comment]
@funyapu Could you create a minimal reproduction for your case?

**Comment by funyapu** [comment]
@Awjin  Please do forget my commend. I just found that is because I still have @import in the index.scss :(

**Comment by Carla-H** [comment]
@Awjin I'm working on the same project as @tuelsch - and there is another problem as well. Should I open a new Ticket for this?

Nested placeholders are not available on a @use when referenced a second time - unless they were not actually used the first time.
The current workaround is to just straight up import all files which contain placeholders.

` --- test.scss
@use "foo";
@use "bar";

--- foo.scss
@use "baz";

.foo {
  @extend %baz;
}

--- bar.scss
@use "baz";

.bar {
  @extend %baz;
}

--- baz.scss
@use "qux";

%baz {
  @extend %qux;
}

--- qux.scss
%qux {
  background-color: green;
}
`

This will also produce the error mentioned above.
It will work if: 
- You comment any of the 'extend' (it even works if you comment the extend in foo.scss)
- You comment any of the 'use' in the test.scss

**Comment by cpof-tea** [comment]
@Carla-H it is recommended to use mixins instead of placeholders

**Comment by Carla-H** [comment]
@cpof-tea We have our reasons to want to use placeholders for certain elements - and even if I used mixins here, the bug itself wouldn't vanish from sass.

**Comment by nikonovsasha** [comment]
Hi @tuelsch , hi SASS team!
First, thank you for a great tool! SASS is beautyful.
I have the same issue, @Carla-H has. I would to show you the real-life example:

shared/header.sass
----------------
%header
    font-size: 10px

shared/footer.sass
---------------
%footer
    background: red

shared/layout.sass
---------------
@use "header"
@use "footer"

%default-layout
    .header
        @extend %header

    .footer
         @extend %footer

single-product-page.sass
---------------------------------
@use "shared/layout"

.single-product-page
    @extend %default-layout

front-page.sass
---------------------
@use "shared/layout"

.front-page
    @extend %default-layout

entry.sass
--------------
@use "single-product-page"
@use "front-page"

I think, this is a great scenario to use SASS. It provides the real modularity and independence for every page type.
I am starting a new project right now and I would like to use this approach, but I just faced the @Carla-H 's issue.
Could you tell me, please - will SASS team fix this issue in the future or would this usage be considered as "wrong"? If yes, then I can use @import as temporary fix and then remove it. If not, then I, obviously, can't use this approach

**Comment by nikonovsasha** [comment]
Hi @Awjin , could you answer the above question, please? I have confused the nicknames and accidentally addressed it to topic starter

**Comment by Awjin** [comment]
Sorry for the delayed response, I'm looking into this.

**Comment by nikonovsasha** [comment]
Hi Awjin, all I need to know is that SASS team is going to support this in the future, so thank you for your answer) Have a good day!

**Comment by nikistefanov** [comment]
Hi SASS team,

I've hit another issue with modules and %placeholder selector. Whenever I `@extend` %selector that is declared in the same file everything works fine, but when I `@use` the file that has the %selector the only thing that is compiled is the root level rules and the ones that are nested are missing. Here is an example of this scenario and also a [code demo repo](https://github.com/nikistefanov/sass-demo).

````scss
placeholders/_link.scss
------------------------

%default-link {
    color: blue;
    
    &.-red {
        color: red;
    }
}

components/_link.scss
------------------------
@use "../placeholders/link";

.a {
  @extend %default-link;
}
````

This will generate only. No `.-red` modifier
````css
main.css
-----------------------
.a {
  color: blue;
}
````

**Comment by tuelsch** [comment]
@Awjin my original issue seems to be resolved with version 1.35.1. @nikistefanov and @nikonovsasha could you check if your issues are resolved as well?

**Comment by nikistefanov** [comment]
@gfellerph yes, now the output contains the modifier. Thanks!

**Comment by jrson83** [comment]
For anyone coming here looking for information on how to `@use/@forward` `%placeholders` with `@mixins`, I wrote a blog post about:

https://jrson.me/blog/how-to-use/forward-placeholders-and-extend-selectors-with-sass/scss-mixins-effectively/
